### PR TITLE
Improved MEX Compilation

### PR DESCRIPTION
**Pull request by jjangsangy**
## Features
-  Autodetect installed MATLAB bin directory with `get_matlab_bin()` function.
- `messenger` is a fully fledged helper module for `pymatbridge`
    - Meaning you can call code from messenger just like any module.

```python
# From Root Directory
>>> import messenger
>>> messenger.get_matlab_bin()
'/Applications/MATLAB_R2014a.app/bin'
```

- `messenger` module is also a callable application, and is a direct replacement for `make.py`.

```python
>>> python -m messenger --help

usage: messenger [-h] [--static] {matlab,octave}

positional arguments:
  {matlab,octave}  target to be built

optional arguments:
  -h, --help       show this help message and exit
  --static         staticly link libzmq
```

### etc...

- Code is more `DRY`
    - `local.cfg` files are deprecated. Now only single authoritative `config.ini` file in root directory.
- Use `get_matlab_env()` for inspecting underlying installed matlab architecture.
- Use `fetch_zmq()` to download `libzmq` source code

### Still not Implemented
- Code for comparing against python, host and matlab architectures.
- Dynamic linking of `libzmq` headers and library using `DYLD_LIBRARY_PATH` `DYLD_FRAMEWORK_PATH`.
- Updated Documentation

### CODE REVIEW

**Comment by jjangsangy** [comment]
First try! Booya!

**Comment by blink1073** [comment]
Pretty awesome job @jjangsangy!  I'll let @isbadawi chime in since he's a fellow builder.

File: messenger/make.py [hunk 1]
```diff
@@ -1,48 +1,47 @@
...
-    cfg = {}
-    for line in lines:
-        if '=' not in line:
-            continue
-        name, path = line.split('=')
-        cfg[name.lower()] = path.strip() or '.'
+    config = os.path.join(os.path.realpath(__file__), 'config.ini')
+    print(config)
+    cfg = ConfigParser()
+    config.read(config)
```
Comment (isbadawi) [inline messenger/make.py]: Should be `cfg.read(config)`. Can't build Octave because of this.

There are two other places where you read `config.ini`. You should probably reuse this function everywhere.

File: messenger/make.py [hunk 2]
```diff
@@ -68,45 +67,34 @@ def build_octave():
...
-    extension = extension.decode('utf-8').rstrip('\r\n')
+    matlab_bin = settings.get_matlab_bin()
+    cfg, host = ConfigParser(), platform.system()
+    cfg.read('config.ini')
+    libzmq = {
+        'zmq_lib': os.path.normpath(cfg.get(host, 'ZMQ_LIB')),
+        'zmq_inc': os.path.normpath(cfg.get(host, 'ZMQ_INC')),
+    }
+
+    extcmd     = '%s' % os.path.join(matlab_bin, "mexext")
```
Comment (isbadawi) [inline messenger/make.py]: There used to be quotes around the command here. I'm pretty sure they're still needed so you should add them back -- if not, you can drop the string interpolation altogether.

File: messenger/settings.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    logger.info('Downloading ZMQ Source Version %i.%i.%i'     % version)
+    url = ("http://download.zeromq.org/zeromq-%i.%i.%i.tar.gz" % version)
+    name = urlretrieve(url, url.rsplit('/')[-1])[0]
+
+    logger.info('Extracting into prefix %s' % prefix)
+    with tarfile.open(name) as tar:
+        tar.extractall(prefix)
+
+    logging.info('Download Complete')
```
Comment (isbadawi) [inline messenger/settings.py]: You use `logging` here but `logger` before that. I would just use `print`.

File: messenger/__main__.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    parser.add_argument(
+        "target",
+        choices=["matlab", "octave"],
+        type=str.lower,
+        help="target to be built"
+    )
+    parser.add_argument(
+        "--static",
+        action="store_true",
+        help="staticly link libzmq"
```
Comment (isbadawi) [inline messenger/__main__.py]: statically

File: messenger/make.py (hunk 1 above)
Comment (isbadawi) [inline messenger/make.py]: Also this used to return a dict. The octave path uses `% get_config()` and this won't work with a `ConfigParser` instance.

File: messenger/settings.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    environment: dictionary
+        Mapping of environment variable[s]
+    """
+    command = ' '.join([matlab, '-e']),
+    process = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE)
+
+    envs    = (line.decode('utf-8').strip() for line in process.stdout)
+    mapping = (env.split('=', maxsplit=1)   for env  in envs)
+
+    return {key:value for key, value in mapping}
```
Comment (isbadawi) [inline messenger/settings.py]: `return dict(mapping)`

File: setup.py [hunk 6]
```diff
@@ -2,18 +2,18 @@
...
-import sys
 import shutil
 import glob
 
 
 # BEFORE importing distutils, remove MANIFEST. distutils doesn't properly
 # update it when the contents of directories change.
-if os.path.exists('MANIFEST'):
-    os.remove('MANIFEST')
+if os.path.exists('MANIFEST'): os.remove('MANIFEST')
```
Comment (isbadawi) [inline setup.py]: Minor, but I don't see the point of changing this to be on one line.

File: messenger/settings.py (hunk 3 above)
Comment (jjangsangy) [inline messenger/settings.py]: Oh good catch. I was wondering why logging was acting strange.

File: messenger/settings.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        'Linux'  : r'/usr/local',
+    }
+
+    if cfg.has_option(host, 'MATLAB_BIN'):
+        matlab = cfg.get(host, 'MATLAB_BIN')
+    else:
+        # Searches for Matlab bin if it's not set
+        _root    = glob(r'%s/MATLAB*' % programs[host])[0]
+        _version = [p for p in os.listdir(_root) if p.startswith('R20')]
+        if any(_version):
```
Comment (isbadawi) [inline messenger/settings.py]: `if _version:`

File: setup.py [hunk 8]
```diff
@@ -43,6 +43,7 @@
             version=VERSION,
             packages=PACKAGES,
             package_data=PACKAGE_DATA,
+            zip_safe=False,
```
Comment (isbadawi) [inline setup.py]: What's the story here?

File: messenger/__main__.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+
+
+def main():
+    args = command_line()
+    if args.target == "matlab":
+        build_matlab(static=args.static)
+    elif args.target == "octave":
+        build_octave()
+    else:
```
Comment (isbadawi) [inline messenger/__main__.py]: Don't need this `else` -- `argparse` will throw an error if the target is something else.

**Comment by isbadawi** [comment]
Lots of little nits, but OK overall. I do like the one config file better.

File: setup.py (hunk 8 above)
Comment (jjangsangy) [inline setup.py]: So we're switching from `distutils.core` to `setuptools` and `setuptools` installs the package in the egg binary format. Setting `zip_safe` to `False` bypasses that whole calculation and installs both the egg and source formats in site-packages.

https://pythonhosted.org/setuptools/setuptools.html#setting-the-zip-safe-flag

One alternative would be using the wheel format.

**Comment by jjangsangy** [comment]
Ya, thanks a lot for going through and finding all those mistakes. Always nice to get code reviewed!

Went ahead and refactored out all the details.

`get_config()` now actually returns a dict for the host platform. I just pass that around now instead of having every function re-calculate those values.

**Comment by jjangsangy** [comment]
Crap! forgot a comma!

File: messenger/make.py [hunk 10]
```diff
@@ -1,112 +1,133 @@
...
-        elif maxint == 2147483647:
-            messenger_dir = 'mexw32'
-    return messenger_dir
-
-
-def get_config():
-    messenger_dir = get_messenger_dir()
-    with open(os.path.join(messenger_dir, 'local.cfg')) as fid:
-        lines = fid.readlines()
+    host, is_64bit = platform.system(), sys.maxsize > 2**32
```
Comment (mmagnuski) [inline messenger/make.py]: This will reintroduce a problem I had running `pymatbridge` with 32 bit python on 64 bit Windows.
If you are running 32 bit python on 64 bit windows `sys.maxsize > 2**32` gives `False` while `platform.machine().endswith('64')` returns `True`.

Comment (jjangsangy) [inline messenger/make.py]: @mmagnuski Thanks for bringing that up!

Planning on dealing with that soon. The issue is that we actually have 3 different architectures.

1. Operating System Architecture
2. Python Architecture
3. Matlab Architecture

I just built `get_matlab_env()` which will return the MATLAB arch, so now all we have to do is figure out which combinations of these 3 are okay to run under.

Comment (blink1073) [inline messenger/make.py]: IMO, the only one that matters for the mex file is the Matlab Architecture, which must in turn match the ZMQ Architecture.

Comment (jjangsangy) [inline messenger/make.py]: But using a 64-bit compiled MEX running on a 32-bit Python interpreter is bound to be bad. I'm not too familiar with the build process for windows so I'll have to go look into this one.

ZMQ Architecture? Or do you mean the version of the ZMQ installation?

Comment (blink1073) [inline messenger/make.py]: My point is that we Matlab it in a subprocess and talk to it over sockets, so it doesn't matter if the architectures match.  
There are 64bit and 32bit downloads for ZMQ on Windows, for example.

Comment (jjangsangy) [inline messenger/make.py]: Ahh I see, that makes sense.

File: setup.py [hunk 11]
```diff
@@ -1,54 +1,71 @@
...
-            scripts=BIN
-            )
 
+for f in glob.glob("messenger/%s/messenger.*" % get_messenger_dir()):
+    shutil.copy(f, 'pymatbridge/matlab/')
 
 # Now call the actual setup function
 if __name__ == '__main__':
-    setup(**opts)
+    setup(
```
Comment (arokem) [inline setup.py]: Why did you move all this here? Seems unrelated to the main point of this PR.

File: messenger/make.py [hunk 12]
```diff
@@ -1,112 +1,133 @@
...
-    elif args.target == "octave":
-        build_octave()
+def get_matlab_bin(config='config.ini'):
+    """
+    Tries to find the MATLAB bin directory independent on host platform.
+    The operation of this function can be overridden by setting the MATLAB_BIN
+    variable within the configuration file specified.
+
+    Parameters
+    -=========
```
Comment (arokem) [inline messenger/make.py]: Please use "------" for underlines here. We try to follow the numpy docstring standard: 

https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#docstring-standard

**Comment by arokem** [comment]
Overall, this looks good to me. Slowly moving towards the day when we don't have to ship the built messengers, but instead build them using the setup machinery. 

A couple of comments/questions: 

#. Do you think it would be possible to write a few free-standing tests for the `messenger` module? It would be nice to test that the functional interface is robust, and to detect bugs, as we keep working on this. 

#. If I understand correctly @mmagnuski's comment, you should probably revert to the checks that were previously done on sys.platform, instead of the 64 bit check you have currently. Is that correct? 

Also - in general, I would prefer to make cleanups on a separate PR, if that's possible, since they're quite unrelated to this, and it just makes reviewing the relevant parts of the code a little bit more difficult. Awesome work!

**Comment by jjangsangy** [comment]
Sure, thing.

How about we close this one, and I'll submit a new PR with all the clean ups discussed. That way, we can at least get something pushed to master without building up too much stuff.

I'm also currently working on more long term packaging related things on another branch, but that can wait for another time!

File: messenger/make.py (hunk 12 above)
Comment (jjangsangy) [inline messenger/make.py]: Cool, I've actually never seen this before. I'll give it a good readthrough!

File: setup.py (hunk 11 above)
Comment (jjangsangy) [inline setup.py]: Sorry, I forgot about these changes, they should be on another brach!

Comment (arokem) [inline setup.py]: Yeah - it's rather well thought-through. In particular, it plays very well
with sphinx, to build nice documentation pages, which is something we
should do one day...

**Comment by arokem** [comment]
And yes - that sounds like an excellent plan. I look forward to see what you come up with. It's good to stay coordinated with the jupyter team. One of the medium-term goals of this work is to have a direct matlab kernel for jupyter.

**Comment by arokem** [comment]
Closing this, in favor of some more fine-grained PRs to follow.

File: setup.py (hunk 11 above)
Comment (jjangsangy) [inline setup.py]: This is something I'm already working on. I can have a reference implementation ready for sphinx-build within a couple of days.

Especially since the documentation is already in reStructuredText, I would really like to get that operable. I'll make sure to put it into another PR we can look into later. I'll can submit a basic scaffold and automated build for getting the docs up on https://readthedocs.org.
